pallets/capacity/src/lib.rs

```diff
@@ -477,6 +715,36 @@ impl<T: Config> Pallet<T> {
 		Ok(capacity)
 	}
 
+	fn increase_stake_and_issue_boost_capacity(
+		staker: &T::AccountId,
+		staking_details: &mut StakingDetails<T>,
+		target: &MessageSourceId,
+		amount: &BalanceOf<T>,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		staking_details.deposit(*amount).ok_or(ArithmeticError::Overflow)?;
+		Self::set_staking_account_and_lock(staker, staking_details)?;
+
+		// get the capacity generated by a Provider Boost
+		let capacity = Self::capacity_generated(T::RewardsProvider::capacity_boost(*amount));
+
+		let mut target_details = Self::get_target_for(staker, target).unwrap_or_default();
+
+		target_details.deposit(*amount, capacity).ok_or(ArithmeticError::Overflow)?;
+		Self::set_target_details_for(staker, *target, target_details);
+
+		let mut capacity_details = Self::get_capacity_for(target).unwrap_or_default();
+		capacity_details.deposit(amount, &capacity).ok_or(ArithmeticError::Overflow)?;
+		Self::set_capacity_for(*target, capacity_details);
+
+		let era = Self::get_current_era().era_index;
+		Self::upsert_boost_history(staker, era, *amount, true)?;
```

Comment:
I believe we should avoid writing upsert functions whenever possible. While they can be useful in certain situations, I think we can take a different approach here. We’ve encountered multiple bugs in our code due to this style, so it might be better to consider alternatives.